core-v3/src/main/java/io/roastedroot/protobuf4j/Protobuf.java

```diff
@@ -210,6 +210,92 @@ public static DescriptorProtos.FileDescriptorSet getDescriptors(
+     * performing semantic validation like type checking across files.
+     *
+     * <p>The validation checks:
+     *
+     * <ul>
+     *   <li>Proto file syntax (syntax version, message structure, field definitions)
+     *   <li>Field numbers, types, and options within the file
+     *   <li>Message and field naming conventions
+     *   <li>Proto grammar compliance
+     * </ul>
+     *
+     * <p>The validation does NOT check:
+     *
+     * <ul>
+     *   <li>Whether imported files exist
+     *   <li>Whether types referenced from imports are defined
+     *   <li>Cross-file type compatibility
+     * </ul>
+     *
+     * @param workdir the working directory containing proto files
+     * @param fileName the proto file name to validate (single file only)
+     * @return ValidationResult indicating success or containing error messages
+     */
+    public static ValidationResult validateSyntax(Path workdir, String fileName) {
+        try (ByteArrayOutputStream stdout = new ByteArrayOutputStream();
+                ByteArrayOutputStream stderr = new ByteArrayOutputStream()) {
+            var wasiOptsBuilder = WasiOptions.builder().withStdout(stdout).withStderr(stderr);
+
+            List<String> command = new ArrayList<>();
+            command.add("protoc-wrapper");
+            command.add("validate-syntax");
+            command.add(fileName);
+
+            var wasiOpts =
+                    wasiOptsBuilder
+                            .withArguments(command)
+                            .withDirectory(workdir.toString(), workdir)
+                            .build();
+            try (var wasi = WasiPreview1.builder().withOptions(wasiOpts).build()) {
+                var imports =
+                        ImportValues.builder()
+                                .addFunction(wasi.toHostFunctions())
+                                .addMemory(defaultMemory())
+                                .build();
+
+                LOGGER.log(
+                        Level.FINE,
+                        "protoc command: " + command.stream().collect(Collectors.joining(" ")));
+                Instance.builder(PROTOBUF_WRAPPER)
+                        .withImportValues(imports)
+                        .withMachineFactory(ProtobufWrapper::create)
+                        .build();
+            } catch (WasiExitException exit) {
+                if (exit.exitCode() != 0) {
+                    // Parse error messages from stderr
+                    String errorOutput = stderr.toString();
+                    if (errorOutput != null && !errorOutput.isEmpty()) {
+                        return ValidationResult.invalid(errorOutput.trim());
+                    }
+                    return ValidationResult.invalid("Validation failed with exit code: " + exit.exitCode());
+                }
+            }
+
+            // Success - check stdout for "OK"
```

Comment:
Isn't this redundant?
We are already checking for the 0 return code a few lines earlier.